stdlib/public/Concurrency/Task.cpp

```diff
@@ -1405,27 +1451,41 @@ static void resumeTaskAfterContinuation(AsyncTask *task,
   // restarting.
   _swift_tsan_release(static_cast<Job *>(task));
 
-  // The status should be either Pending or Awaited.  If it's Awaited,
-  // which is probably the most likely option, then we should immediately
-  // enqueue; we don't need to update the state because there shouldn't
-  // be a racing attempt to resume the continuation.  If it's Pending,
-  // we need to set it to Resumed; if that fails (with a strong cmpxchg),
-  // it should be because the original thread concurrently set it to
-  // Awaited, and so we need to enqueue.
+  // The status should be either Pending or Awaited.
+  //
+  // Case 1: Status is Pending
+  // No one has awaited us, we just need to set it to Resumed; if that fails
+  // (with a strong cmpxchg), it should be because the original thread
+  // concurrently set it to Awaited, in which case, we fall into Case 2.
+  //
+  // Case 2: Status is Awaited
+  // This is probably the more frequently hit case.
+  // In task-to-thread model, we update status to be Resumed and signal the
+  // waiting thread. In regular model, we immediately enqueue the task and can
+  // skip updates to the continuation state since there shouldn't be a racing
+  // attempt to resume the continuation.
   if (status == ContinuationStatus::Pending &&
       sync.compare_exchange_strong(status, ContinuationStatus::Resumed,
                                    /*success*/ std::memory_order_release,
-                                   /*failure*/ std::memory_order_relaxed)) {
+                                   /*failure*/ std::memory_order_acquire)) {
     return;
   }
   assert(status == ContinuationStatus::Awaited &&
          "detected concurrent attempt to resume continuation");
+#if SWIFT_CONCURRENCY_TASK_TO_THREAD_MODEL
+  // If we see status == ContinuationStatus::Awaited, then we should also be
+  // seeing a pointer to the cond var since we're doing a load acquire on sync
+  // which pairs with the store release in swift_continuation_awaitImpl
+  assert(context->Cond != nullptr);
 
+  sync.store(ContinuationStatus::Resumed, std::memory_order_relaxed);
```

Comment:
It's _probably_ OK, but I'm a little concerned that this should be `std::memory_order_release` and the one in the condition variable loop should be `std::memory_order_acquire`. If the variable was actually protected here by the lock, I think relaxed would be _fine_, but because it isn't I just worry that there's a potential deadlock (where we wake up, but don't see the `Resumed` status).

Reply:
Hmmm, we're not trying to make sure that a write to a different memory location is also visible as a result of this store here - it's all self-contained in the atomic. 

If a thread wakes up from a signal() and doesn't see that this write has happened, then I would say that's a bug in the cond var's signal implementation missing a release barrier inside it?

Reply:
I just worry that someone might decide that the barriers belong in the lock implementation rather than in the signal implementation. I'm probably worrying unnecessarily, it's probably fine as I said.

Reply:
Yea I think this should be just fine.

> I just worry that someone might decide that the barriers belong in the lock implementation rather than in the signal implementation. 

I think such an implementation of cond vars would be invalid since most of them don't necessitate that the caller hold the associated mutex when they signal and therefore, they have to have the right barriers.